src/crab/ebpf_domain.hpp

```diff
@@ -521,6 +527,21 @@ class ebpf_domain_t final {
         if (!when_stack.is_bottom()) {
             if (!stack.all_num(when_stack, lb, ub)) {
                 require(when_stack, access_reg.type != T_STACK, "Illegal map update with a non-numerical value.");
+            } else if (thread_local_options.strict && map_type.has_value() && map_type->is_array) {
+                // Get offset value.
+                variable_t key_ptr = access_reg.offset;
+                std::optional<number_t> offset = m_inv[key_ptr].singleton();
+                if (!offset.has_value()) {
+                    // REVIEW: is there a better way to fail validation here?
+                    require(m_inv, linear_expression_t(0) != 0, "Pointer must be a singleton");
+                } else if (s.key) {
+                    // Look up the value pointed to by the key pointer.
+                    variable_t key_value =
+                        variable_t::cell_var(data_kind_t::values, (uint64_t)offset.value(), sizeof(uint32_t));
```

Comment:
This too might be useful more generally so it merits a method, though I don't think it's needed anywhere else right now.